spock-core/src/main/java/org/spockframework/runtime/DataIteratorFactory.java

```diff
@@ -319,42 +468,302 @@ private Object[] getPreviousDataTableProviders(List<String> dataProviderVariable
+                                DataProviderInfo providerInfo, Object provider) {
+      super(supervisor, context);
+      this.dataVariableNames = singletonList(Objects.requireNonNull(dataVariableName));
+      estimatedNumIterations = estimateNumIterations(Objects.requireNonNull(provider));
+      this.providerInfo = Objects.requireNonNull(providerInfo);
+      this.provider = provider;
+      iterator = createIterator(provider, providerInfo);
+    }
+
+    @Override
+    public void close() throws Exception {
+      GroovyRuntimeUtil.invokeMethodQuietly(provider, "close");
+    }
+
+    @Override
+    public boolean hasNext() {
+      if (context.getErrorInfoCollector().hasErrors()) {
+        return false;
+      }
+
+      try {
+        return iterator.hasNext();
+      } catch (Throwable t) {
+        supervisor.error(context.getErrorInfoCollector(), new ErrorInfo(providerInfo.getDataProviderMethod(), t));
+        return false;
+      }
+    }
 
-    private Iterator<?>[] createIterators() {
+    @Override
+    public Object[] next() {
       if (context.getErrorInfoCollector().hasErrors()) {
         return null;
       }
 
-      Iterator<?>[] iterators = new Iterator<?>[dataProviders.length];
-      for (int i = 0; i < dataProviders.length; i++)
-        try {
-          Iterator<?> iter = GroovyRuntimeUtil.asIterator(dataProviders[i]);
-          if (iter == null) {
-            supervisor.error(context.getErrorInfoCollector(), new ErrorInfo(context.getCurrentFeature().getDataProviders().get(i).getDataProviderMethod(),
-              new SpockExecutionException("Data provider's iterator() method returned null")));
+      try {
+        return new Object[]{iterator.next()};
+      } catch (Throwable t) {
+        supervisor.error(context.getErrorInfoCollector(), new ErrorInfo(providerInfo.getDataProviderMethod(), t));
+        return null;
+      }
+    }
+
+    @Override
+    public int getEstimatedNumIterations() {
+      return estimatedNumIterations;
+    }
+
+    @Override
+    public List<String> getDataVariableNames() {
+      return dataVariableNames;
+    }
+  }
+
+  private static class DataProviderMultiplier extends BaseDataIterator {
+    private final List<String> dataVariableNames;
+    private final List<DataProviderInfo> multiplierProviderInfos;
+    private final List<DataProviderInfo> multiplicandProviderInfos;
+    private final DataProviderMultiplier multiplierProvider;
+    private final Object[] multiplierProviders;
+    private final DataProviderMultiplier multiplicandProvider;
+    private final Object[] multiplicandProviders;
+    private final Iterator<?>[] multiplierIterators;
+    private final Iterator<?>[] multiplicandIterators;
+    private boolean factorsSwapped = false;
+    private Object[] currentMultiplierValues = null;
+    private boolean collectMultiplicandValues = true;
+    private final List<List<Object>> collectedMultiplicandValues;
+    private final int estimatedNumIterations;
```

Comment:
I don't think this will add too much clarity, because practically all of these are nullable in one situation or the other. A data provider multiplier handles the multiplication of either a set of data providers with another set of data providers (`(a * b)` in `((a * b) * c)`), or of one data provider multiplication with a set of data providers (`(ab * c)` in `((a * b) * c)`). And in both cases the factors can be swapped around to store as few values as possible in memory. Due to this all could be `null` depending on concrete situation.

But I added some significantly more amount of docs about it to the fields and constructors, that hopefully clears it up enough.